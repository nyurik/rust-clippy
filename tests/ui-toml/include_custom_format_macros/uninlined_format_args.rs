#![warn(clippy::uninlined_format_args)]
#![allow(named_arguments_used_positionally, unused_imports, unused_macros, unused_variables)]
#![allow(clippy::eq_op, clippy::format_in_format_args, clippy::print_literal)]

macro_rules! _internal {
    ($($args:tt)*) => {
        println!("{}", format_args!($($args)*))
    };
}

macro_rules! my_println2 {
   ($target:expr, $($args:tt)+) => {{
       if $target {
           _internal!($($args)+)
       }
    }};
}

macro_rules! my_println2_args {
    ($target:expr, $($args:tt)+) => {{
       if $target {
           _internal!("foo: {}", format_args!($($args)+))
       }
    }};
}

macro_rules! my_concat {
    ($fmt:literal $(, $e:expr)*) => {
        println!(concat!("ERROR: ", $fmt), $($e,)*)
    }
}

macro_rules! my_good_macro {
    ($fmt:literal $(, $e:expr)* $(,)?) => {
        println!($fmt $(, $e)*)
    }
}

macro_rules! my_bad_macro {
    ($fmt:literal, $($e:expr),*) => {
        println!($fmt, $($e,)*)
    }
}

macro_rules! my_bad_macro2 {
    ($fmt:literal) => {
        let s = $fmt.clone();
        println!("{}", s);
    };
    ($fmt:literal, $($e:expr)+) => {
        println!($fmt, $($e,)*)
    };
}

// This abomination was suggested by @Alexendoo, may the Rust gods have mercy on their soul...
// https://github.com/rust-lang/rust-clippy/pull/9948#issuecomment-1327965962
macro_rules! used_twice {
    (
        large = $large:literal,
        small = $small:literal,
        $val:expr,
    ) => {
        if $val < 5 {
            println!($small, $val);
        } else {
            println!($large, $val);
        }
    };
}

fn main() {
    let local_i32 = 1;
    println!("val='{}'", local_i32);
    my_println2_args!(true, "{}", local_i32);
    my_println2!(true, "{}", local_i32);
    my_concat!("{}", local_i32);
    my_good_macro!("{}", local_i32);
    my_good_macro!("{}", local_i32,);

    // FIXME: Broken false positives, currently unhandled
    // my_bad_macro!("{}", local_i32);
    // my_bad_macro2!("{}", local_i32);
    // used_twice! {
    //     large = "large value: {}",
    //     small = "small value: {}",
    //     local_i32,
    // };
}
