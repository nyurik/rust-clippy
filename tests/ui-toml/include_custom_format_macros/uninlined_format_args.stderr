error: variables can be used directly in the `format!` string
  --> tests/ui-toml/include_custom_format_macros/uninlined_format_args.rs:73:5
   |
LL |     println!("val='{}'", local_i32);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `-D clippy::uninlined-format-args` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::uninlined_format_args)]`
help: change this to
   |
LL -     println!("val='{}'", local_i32);
LL +     println!("val='{local_i32}'");
   |

error: variables can be used directly in the `format!` string
  --> tests/ui-toml/include_custom_format_macros/uninlined_format_args.rs:77:5
   |
LL |     my_good_macro!("{}", local_i32);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: change this to
   |
LL -     my_good_macro!("{}", local_i32);
LL +     my_good_macro!("{local_i32}");
   |

error: variables can be used directly in the `format!` string
  --> tests/ui-toml/include_custom_format_macros/uninlined_format_args.rs:78:5
   |
LL |     my_good_macro!("{}", local_i32,);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: change this to
   |
LL -     my_good_macro!("{}", local_i32,);
LL +     my_good_macro!("{local_i32}",);
   |

error: aborting due to 3 previous errors
